dev/main.go

```diff
@@ -25,16 +27,22 @@ func New(
 	// +optional
 	version string,
 	// +optional
+	gitBranch string,
+	// +optional
+	tagSuffix string,
+	// +optional
 	dockerCfg *Secret,
 ) (*DaggerDev, error) {
-	versionInfo, err := newVersion(ctx, source, version)
+	versionInfo, err := newVersion(ctx, source, version, tagSuffix)
```

Comment:
The tag suffix shouldn't go in here - I think it only makes sense to pass it to `Publish`.

---

Note that by passing it into `DaggerDev.Version`, this means that binaries built inside this engine, like the dagger binary will have `-gpu` in the version name, so we'll get versions of the form `v0.12.0-gpu` which will be evaluated as a prerelease of `v0.12.0`, which isn't the right intention.

We shouldn't build the tag suffix into the binaries, so it probably shouldn't go here. If we *absolute really* want to do this, we could put `+gpu` so it's a build variant instead (but I don't actually think it should go there at all).

Reply:
That's the behaviour that we have today. In the last release, we published:
- `registry.dagger.io/engine:v0.11.9`
- `registry.dagger.io/engine:v0.11.9-gpu`

While changing to `registry.dagger.io/engine:v0.12.0+gpu` makes sense, more changes are coming in the PR that builds on top of this: https://github.com/dagger/dagger/pull/7782

If all version info is collected here, why would we move the tag suffix into `Publish`?

Does this change build the tag suffix into the CLI binary? My understanding is that this only affects the Engine, not the CLI. If you are sure that this will affect the CLI binary, I will set everything up in my fork to test publishing from `main` before we merge this.

Reply:
This isn't quite the behavior we have today - the tags are the same, yes, but the contents of the published image is now different.

To see this:

```
❯ dagger -m dev call --source=.:default --git-branch=main --tag-suffix=bad --version=v0.12.0 dev terminal

dagger /mnt $ dagger version
dagger v0.12.0-bad (registry.dagger.io/engine) linux/amd64
dagger /mnt $ echo "{version}" | dagger query
{
    "version": "v0.12.0-bad"
}
```

The suffix *must not* appear in the version info here - there's some discussion in #7760 around populating these, but it's still a bit unclear. If we don't change this, then this build, despite being of v0.12.0 will look like a pre-release of v0.12.0 - which will break a lot of the work we're version compat (e.g. dagger v0.12.0 clients will entirely reject all connections to v0.12.0-gpu, since v0.12.0-gpu < v0.12.0 :scream:). Unfortunately, since we don't have gpu/wolfi variant provisioning tests, it's unlikely we would have caught this in tests once merged.

These versions are inserted at this point in the builder:

https://github.com/dagger/dagger/blob/2f9e3b39504b7a2798ca24ede94974d9b1e8e437/dev/internal/build/builder.go#L313-L315

The confusion is coming from the changes that now treat the `Version` struct for two separate purposes - determining all the refs/tags to tag with, and determining the version info in the builder, these are different, and need to be separate.

Splitting out the `taxSuffix` to just be on publish would hopefully make this issue much less likely to happen - implementing this comment here would completely resolve this https://github.com/dagger/dagger/pull/7781#discussion_r1658921778, since it's much more explicit.

Reply:
That was a good point. I started from scratch and reworked the implementation. WDYT?